File: advanced/segmented-sieve-of-eratosthenes/segmented-sieve.py
import math

def simpleSieve(n):
  result = []
  primes = [True for _ in range(n + 1)]
  primes[0] = False
  primes[1] = False

  indexI = 2

  while indexI <= math.sqrt(n):
    if primes[indexI]:
      multiplesOfIndexI = indexI ** 2
      indexJ = multiplesOfIndexI

      while indexJ <= n:
        primes[indexJ] = False
        indexJ += indexI
    indexI += 1

  for ind in range(2, n):
    if primes[ind]:
      result.append(ind)

  return result

def segmentedSieve(n):

  # add 1 to range so that very small numbers <= 10 would play nicely
  # for the first prime
  segmentRange = math.floor(math.sqrt(n)) + 1

  firstPrimeSlice = simpleSieve(segmentRange)
  primes = [p for p in firstPrimeSlice]
  result = [p for p in firstPrimeSlice]

  lowerLimit = segmentRange
  upperLimit = segmentRange * 2

  while lowerLimit < n:
    if upperLimit > n:
      upperLimit = n

    currentArray = [True for _ in range(segmentRange)]
    for indI in range(len(primes)):
      prime = primes[indI]

      # find minimum multiple of prime (aka, divisible by prime) in current range
      minMulOfPrime = int(math.floor(lowerLimit / prime) * prime)

      if minMulOfPrime < lowerLimit:
        # keep minimum multiple in current range by shifting to the next multiple in range
        minMulOfPrime += prime

      # mark multiples of prime in range

      for indJ in range(minMulOfPrime, upperLimit, prime):
        indexInRange = indJ - lowerLimit
        currentArray[indexInRange] = False

    for ind in range(lowerLimit, upperLimit):
      # since currentArray is a slice of length low - high,
      # get number index in range by subtractinng index from low
      numIndexInRange = ind - lowerLimit
      if currentArray[numIndexInRange] == True:
        result.append(ind)
      
    # update upper and lower limits for next segment
    lowerLimit += segmentRange
    upperLimit += segmentRange

  return result


print(segmentedSieve(30))

File: advanced/segmented-sieve-of-eratosthenes/segmented-sieve.js
function simpleSieve(n) {
  const primes = new Int8Array(n);
  const result = [];
  primes[0] = 1;
  primes[1] = 1;

  for (let i = 2; i <= Math.sqrt(n); i++) {
    if (primes[i] === 0) {
      const multiplesOfI = i ** 2;

      for (let j = multiplesOfI; j <= n; j += i) {
        primes[j] = 1;
      }
    }
  }

  for (let i = 2; i <= primes.length; i++) {
    if (primes[i] === 0) {
      result.push(i);
    }
  }

  return result;
}

function segmentedSieve(n) {
  const primes = [];
  const result = [];

  // add 1 to range so that very small numbers <= 10 would play nicely
  // for the first prime
  const range = Math.floor(Math.sqrt(n)) + 1;

  const firstPrimesSlice = simpleSieve(range);

  primes.push(...firstPrimesSlice);
  result.push(...firstPrimesSlice);

  let lowerLimit = range;
  let upperLimit = range * 2;

  while (lowerLimit < n) {
    if (upperLimit > n) {
      upperLimit = n;
    }

    const currentArray = new Int8Array(range);

    for (let i = 0; i < primes.length; i++) {
      const prime = primes[i];

      // find minimum multiple of prime (aka, divisible by prime) in current range
      let minMulOfPrime = Math.floor(lowerLimit / prime) * prime;

      if (minMulOfPrime < lowerLimit) {
        // keep minimum multiple in current range by shifting to the next multiple in range
        minMulOfPrime += prime;
      }

      // mark multiples of prime in range
      for (let j = minMulOfPrime; j <= upperLimit; j += prime) {
        const indexInRange = j - lowerLimit;
        currentArray[indexInRange] = 1;
      }
    }

    for (let i = lowerLimit; i <= upperLimit; i++) {
      // since currentArray is a slice of length low - high,
      // get number index in range by subtractinng index from low
      const numIndexInRange = i - lowerLimit;
      if (currentArray[numIndexInRange] === 0) {
        result.push(i);
      }
    }

    // update upper and lower limits for next segment
    lowerLimit += range;
    upperLimit += range;
  }

  return result;
}

console.log(segmentedSieve(30));
